pallets/xyk/src/lib.rs

```diff
@@ -11,92 +8,285 @@ mod mock;
     #[pallet::event]
     #[pallet::generate_deposit(pub(super) fn deposit_event)]
     pub enum Event<T: Config> {
-        /// Event documentation should end with an array that provides descriptive names for event
-        /// parameters. [something, who]
-        SomethingStored(u32, T::AccountId),
+        Swapped {
+            who: T::AccountId,
+            from: (AssetId<T>, Balance<T>),
+            to: (AssetId<T>, Balance<T>),
+        },
+        Locked {
+            who: T::AccountId,
+            x: (AssetId<T>, Balance<T>),
+            y: (AssetId<T>, Balance<T>),
+            lp: (AssetId<T>, Balance<T>),
+        },
+        Unlocked {
+            who: T::AccountId,
+            x: (AssetId<T>, Balance<T>),
+            y: (AssetId<T>, Balance<T>),
+            lp: (AssetId<T>, Balance<T>),
+        },
+        LPMinted {
+            who: T::AccountId,
+            x: (AssetId<T>, Balance<T>),
+            y: (AssetId<T>, Balance<T>),
+            lp: (AssetId<T>, Balance<T>),
+        },
     }
 
     // Errors inform users that something went wrong.
     #[pallet::error]
     pub enum Error<T> {
-        /// Error names should be descriptive.
-        NoneValue,
-        /// Errors should have helpful documentation associated with them.
-        StorageOverflow,
+        /// No liquidity available
+        NoLiquidity,
+        /// Not enough tokens
+        NotEnoughTokens,
+        LiquidityAlreadyExists,
+    }
+
+    impl<T: Config> Pallet<T> {
+        fn transfer(
+            from: T::AccountId,
+            to: T::AccountId,
+            asset_id: AssetId<T>,
+            amount: Balance<T>,
+        ) -> Result<(), Error<T>> {
+            pallet_assets::Pallet::<T>::transfer(
+                RawOrigin::Signed(from).into(),
+                asset_id.into(),
+                T::Lookup::unlookup(to),
+                amount,
+            )
+            .map_err(|_| Error::<T>::NotEnoughTokens)
+        }
+
+        fn mint_lp(lp: AssetId<T>, amount: Balance<T>, to: T::AccountId) {
+            pallet_assets::Pallet::<T>::mint(
+                RawOrigin::Signed(T::DEXAddr::get()).into(),
+                lp.into(),
+                T::Lookup::unlookup(to),
+                amount,
+            )
+            .unwrap()
+        }
+
+        fn burn_lp(lp: AssetId<T>, amount: Balance<T>, from: T::AccountId) {
+            pallet_assets::Pallet::<T>::burn(
+                RawOrigin::Signed(T::DEXAddr::get()).into(),
+                lp.into(),
+                T::Lookup::unlookup(from),
+                amount,
+            )
+            .unwrap()
+        }
+
+        pub fn lp_token(x: AssetId<T>, y: AssetId<T>) -> AssetId<T> {
+            let (x, y) = if x > y { (y, x) } else { (x, y) };
+            AssetId::<T>::decode(&mut T::AssetIdHash::hash_of(&(x, y)).encode().as_ref()).unwrap()
+        }
+
+        pub fn _swap(
+            origin: T::AccountId,
+            (from_id, from): (AssetId<T>, Balance<T>),
```

Comment:
```suggestion
            (from_id, from_balance): (AssetId<T>, Balance<T>),
```

It is always better to be explicit. IDE typically autocompletes variables anyway, so focus on readability instead of short variable names.